test/test_iterdatapipe.py

```diff
@@ -1619,6 +1621,280 @@ def _helper(input_data, exp_res, async_fn, input_col=None, output_col=None, max_
+
+        # this doesn't work atm
+        # __len__ Test: inherits length from source DataPipe
+        # self.assertEqual(target_length, len(map_dp))
+
+        input_dp_nl = IDP_NoLen(range(target_length))
+        map_dp_nl = input_dp_nl.thread_map_batches((lambda x: x), batch_size)
+        for x, y in zip(map_dp_nl, range(target_length)):
+            self.assertEqual(x, torch.tensor(y, dtype=torch.float))
+
+        # this doesn't work atm
+        # __len__ Test: inherits length from source DataPipe - raises error when invalid
+        # with self.assertRaisesRegex(TypeError, r"instance doesn't have valid length$"):
+        #     len(map_dp_nl)
+
+        # Test: two independent ThreadPoolExecutors running at the same time
+        map_dp_parallel = input_dp_parallel.thread_map_batches(fn, batch_size)
+        for x, y, z in zip(map_dp, map_dp_parallel, range(target_length)):
+            self.assertEqual(x, torch.tensor(z, dtype=torch.float))
+            self.assertEqual(y, torch.tensor(z, dtype=torch.float))
+
+        # Reset Test: DataPipe resets properly
+        n_elements_before_reset = 5
+        res_before_reset, res_after_reset = reset_after_n_next_calls(map_dp, n_elements_before_reset)
+        self.assertEqual(list(range(n_elements_before_reset)), res_before_reset)
+        self.assertEqual(list(range(target_length)), res_after_reset)
+
+    @suppress_warnings  # Suppress warning for lambda fn
+    def test_threadpool_map_batches_tuple_list_with_col_iterdatapipe(self):
+        batch_size = 3
+
+        def fn_11(d):
+            return -d
+
+        def fn_1n(d):
+            return -d, d
+
+        def fn_n1(d0, d1):
+            return d0 + d1
+
+        def fn_nn(d0, d1):
+            return -d0, -d1, d0 + d1
+
+        def fn_n1_def(d0, d1=1):
+            return d0 + d1
+
+        def fn_n1_kwargs(d0, d1, **kwargs):
+            return d0 + d1
+
+        def fn_n1_pos(d0, d1, *args):
+            return d0 + d1
+
+        def fn_n1_sep_pos(d0, *args, d1):
+            return d0 + d1
+
+        def fn_cmplx(d0, d1=1, *args, d2, **kwargs):
+            return d0 + d1
+
+        p_fn_n1 = partial(fn_n1, d1=1)
+        p_fn_cmplx = partial(fn_cmplx, d2=2)
+
+        def _helper(ref_fn, fn, input_col=None, output_col=None, error=None):
+            for constr in (list, tuple):
+                datapipe = IterableWrapper([constr((0, 1, 2)), constr((3, 4, 5)), constr((6, 7, 8))])
+                if ref_fn is None:
+                    with self.assertRaises(error):
+                        res_dp = datapipe.thread_map_batches(fn, batch_size, input_col, output_col)
+                        list(res_dp)
+                else:
+                    res_dp = datapipe.thread_map_batches(fn, batch_size, input_col, output_col)
+                    ref_dp = datapipe.map(ref_fn)
+                    if constr is list:
+                        ref_dp = ref_dp.map(list)
```

Comment:
In order for this test (copied from `test_map_tuple_list_with_col_iterdatapipe`) to work I need to convert the tuple produced by the `ref_fn` to a list.
Frankly I do not understand why the test for `map `supposedly doesn't fail in Core, when I copy it here it understandbly fails due to the aforementioned reason

Reply:
Do you know which function is failing?

Reply:
> Noob question: Is the [test_map_tuple_list_with_col_iterdatapipe](https://github.com/SvenDS9/pytorch/blob/05943712a443138497c185405b575043b2916f34/test/test_datapipe.py#L1227) actually executed? When I run it locally it fails due to mismatched types. E.g.:
> 
> ```python
> _helper(lambda data: (data[0], data[1], data[0] + data[1]), fn_n1_def, [0, 1], 2)
> ```
> 
The very first test where the return value of the lambda function is always a tuple. Here this is in Line 1722
```
Expected :[[0, 1, 1], [3, 4, 7], [6, 7, 13]]
Actual   :[(0, 1, 1), (3, 4, 7), (6, 7, 13)]
```

Reply:
Ah! I see, thank you for clearing this up!